src/Form/DataTransformerResolver.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\AdminBundle\Form;
+
+use Sonata\AdminBundle\Admin\FieldDescriptionInterface;
+use Sonata\AdminBundle\Form\DataTransformer\BooleanToStringTransformer;
+use Sonata\AdminBundle\Form\DataTransformer\ModelToIdTransformer;
+use Sonata\AdminBundle\Model\ModelManagerInterface;
+use Symfony\Component\Form\DataTransformerInterface;
+use Symfony\Component\Form\Extension\Core\DataTransformer\DateTimeToStringTransformer;
+
+/**
+ * @author Peter Gribanov <info@peter-gribanov.ru>
+ */
+final class DataTransformerResolver
+{
+    /**
+     * @var array<string, DataTransformerInterface>
+     */
+    private $globalCustomTransformers = [];
+
+    /**
+     * @param array<string, DataTransformerInterface> $customGlobalTransformers
+     */
+    public function __construct(array $customGlobalTransformers = [])
+    {
+        foreach ($customGlobalTransformers as $fieldType => $dataTransformer) {
+            $this->addCustomGlobalTransformer($fieldType, $dataTransformer);
+        }
+    }
+
+    public function addCustomGlobalTransformer(string $fieldType, DataTransformerInterface $dataTransformer): void
+    {
+        $this->globalCustomTransformers[$fieldType] = $dataTransformer;
+    }
+
+    public function resolve(
+        FieldDescriptionInterface $fieldDescription,
+        ModelManagerInterface $modelManager
+    ): ?DataTransformerInterface {
+        $dataTransformer = $fieldDescription->getOption('data_transformer');
+
+        // allow override predefined transformers for 'date', 'boolean' and 'choice' field types
+        if ($dataTransformer instanceof DataTransformerInterface) {
+            return $dataTransformer;
+        }
+
+        $fieldType = (string) $fieldDescription->getType();
+
+        // allow override predefined transformers on a global level
+        if (\array_key_exists($fieldType, $this->globalCustomTransformers)) {
+            return $this->globalCustomTransformers[$fieldType];
+        }
+
+        // Handle date type has setter expect a DateTime object
+        if ('date' === $fieldType) {
+            $this->globalCustomTransformers[$fieldType] = new DateTimeToStringTransformer(null, null, 'Y-m-d');
```

Comment:
Is there a reason why we need a new instance every time a date field is rendered?

Reply:
(same for the other transformers)

Reply:
Here we are talking about solving problem #5693. As part of this problem, a translator cannot be created more than once. In addition, we are doing lazy loading ([as i already said](https://github.com/sonata-project/SonataAdminBundle/pull/5937#discussion_r428080908)) and the transformer cannot be initialized more than once.

Or are you talking about [these lines](https://github.com/sonata-project/SonataAdminBundle/blob/3.x/src/Action/SetObjectFieldValueAction.php#L116-L141) that are moved to where they should be, in transformers?

Reply:
> Here we are talking about solving problem #5693. As part of this problem, a translator cannot be created more than once. In addition, we are doing lazy loading (as i already said) and the transformer cannot be initialized more than once.

Also the transformer is only created once, creating new instances is kind of a anti pattern. Code becomes harder to maintain and unit testing is nearly impossible for this fragments. If you overuse this transformer you will end up with multiple instances of the same class that won't differ in any case.

To solve this issue, you could define each transformer as a symfony service and inject it via `addCustomGlobalTransformer`. You could still use lazy bindings for symfony services. This will also solve the testability of this class.

Reply:
> Also the transformer is only created once, creating new instances is kind of a anti pattern. Code becomes harder to maintain and unit testing is nearly impossible for this fragments. If you overuse this transformer you will end up with multiple instances of the same class that won't differ in any case.

@core23 maybe i missed something. How long have Creational design patterns been considered antipatterns?

Reply:
Sorry, I think I misunderstood this class when reviewing it. It is okay to create new instances of a type as long as you use a factory for it. This class can be considered as a factory.

But I would still prefer services over static creation. This will reduce memory footprint and increase testability.

Reply:
Hi, I'm not sure to understand. Do you still request changes or is this PR ok for you @core23 ?

Reply:
After merging #6144, #6167 PRs, i will need to modify this PR. As a result, as a service it will be possible to register only `BooleanToStringTransformer`.

Reply:
I register a `BooleanToStringTransformer` as a service.